Add support for Node.js unofficials builds.

ollliegits opened the pull request:
Added flags `--help`, `--list-unofficial`, `--unofficial <version>` in order to support installation of unofficial builds from nodejs.org.

Commit:
feat: Added option to query and install node from nodejs.org unofficial builds.

Commit:
docs: Readme updated.

Commit:
fix: Spelling of flag 'list-unofficial-releases' corrected to match help text.

Commit:
docs: Readme adapted to changes.

ollliegits committed:
Merge pull request #1 from ollliegits/add-armv6-unofficials

Add support for unofficials releases from nodejs.org

Commit:
feat: Working around required jq by using tail + cut. Additionally checking for required tools.

ollliegits committed:
Merge pull request #2 from ollliegits/add-armv6-unofficials

feat: Working around required jq by using tail + cut. Additionally ch…

MichaIng commented:
Many thanks for your pull request. Sorry that I seem to have missed it so long. I'll have a look over it tomorrow.

MichaIng commented on README.md:
```diff
@@ -1,21 +1,31 @@
 # Node.js Linux Installer
 This is a universal Node.js installer for Linux. The [original project by taaem](https://github.com/taaem/nodejs-linux-installer) required some fixes to maintain functionality, which are contained in this fork, together with a few enhancements.
 
-[![CodeFactor](https://www.codefactor.io/repository/github/michaing/nodejs-linux-installer/badge)](https://www.codefactor.io/repository/github/michaing/nodejs-linux-installer)
```
Please keep the CodeFactor badge in my repo for now. I still use it, although will switch to GitHub action based shellcheck by times.

MichaIng commented on README.md:
```diff
@@ -1,21 +1,31 @@
 # Node.js Linux Installer
 This is a universal Node.js installer for Linux. The [original project by taaem](https://github.com/taaem/nodejs-linux-installer) required some fixes to maintain functionality, which are contained in this fork, together with a few enhancements.
 
-[![CodeFactor](https://www.codefactor.io/repository/github/michaing/nodejs-linux-installer/badge)](https://www.codefactor.io/repository/github/michaing/nodejs-linux-installer)
```
Please keep the CodeFactor badge in my repo for now. I still use it, although will switch to GitHub action based shellcheck by times.

ollliegits replied:
Fixed.

MichaIng commented on README.md:
```diff
@@ -1,21 +1,31 @@
+
+Get the help text:
+
+    curl -sSf https://raw.githubusercontent.com/ollliegits/nodejs-linux-installer/master/node-install.sh | bash -s -- --help
```
Before merging to my repo, these links would need to be changed, of course 😉.

ollliegits replied:
Fixed.

MichaIng commented on node-install.sh:
```diff
@@ -1,6 +1,59 @@
+    -lu
+    --list-unofficial-releases          List all available versions in nodejs.org's unofficial builds.
+    
+    -u <version>
```
Would be worth an own PR, but would like to here your opinion on this:

What do you think about a "latest" option (or making this the default when no version is given), where we use numeric sorting to get the latest version automatically? But:
- Not all versions support all architectures.
- Sorting with v15.9.X and v15.14.X correctly is difficult, without quite some effort to differentiate between major and minor version.

Since the official download page has the same `index.tab`, we could use the same method for both build sources and use the contained architecture list (field 3) to sort out result that do not match the architecture in the first place.

The option could be finally about whether to include the unofficial builds or not and whether to pull a specific version or the latest that exist in either the official or both sources for the own architecture.

ollliegits replied:
I really like the idea of using latest as default, when no version is given.
Is this, amongst others, what you refer to by adding some more command line options? Otherwise I can offer to work on resolving the latest unofficial version for the given architecture. That's if you can afford to wait the extra time if I write the code instead of you.

MichaIng replied:
I just recognised the `sort -V` flag, which makes obtaining the latest version from the list trivial 👍.

I would merge the PR as it is now (after testing the next hours) and we can implement further enhancements with a separate one. To keep it as simple and intuitive as possible:

- Install latest official build available for the architecture by default. But switch to the cleaner method via `index.tab` for official builds as well.
- Use `-V | --version` (or `-v`, but more conflicting with standard arguments?) to optionally define a specific version, rather than the latest.
- Use `-u | --unofficial` to include unofficial builds. We can scrape both `index.tab` lists, sorting out the lines where field 3 does not contain the own architecture, compare and use the newest respectively matching version of both lists. If in rare circumstances both lists contain the same version for the own architecture (probably the case for older `armv6l` versions), the official build is preferred.
- Use `-l | --list` to list all versions available for the own architecture, combine with `-u` (including `-lu`,`-ul` and `--list --unofficial`) to print both lists, for unofficial builds as well. The function to scrape and print the lists can be used to get the latest version as well. When printing to console, we probably find a way to nicely print the lists side by side or one line (+ another line for unofficial builds) to the terminal, to avoid extensive scrolling but have the the latest version at the top-left?

MichaIng replied:
Lol and now I see your commit 😄.

MichaIng commented on node-install.sh:
```diff
@@ -1,6 +1,59 @@
+list_available() {
+    if command -v 'curl' &> /dev/null && command -v 'sort' &> /dev/null; then
+        if command -v 'tail' &> /dev/null && command -v 'cut' &> /dev/null; then
+            curl -l "https://unofficial-builds.nodejs.org/download/release/index.tab" 2>>/dev/null | tail -n +2 | cut -f1 | sort -Vr
```
Little coding hints:
- `command 'xzy' > /dev/null` is better here, as it suppresses any expected output (always STDOUT) but, if it even has an error output, we'd definitely like to know what and which broken bash does this. So keeping STDERR for debugging output that should never happen 😉.
- Quote literal strings with single quotes in bash. In double quotes, variables and command substitutions are expended, which at least marginally slows down processing.
- Below in `[[ "$var" == ... ]]`, the left sided quotation has absolutely zero effect. Only in single angle brackets (test command alias), it prevents word splitting and globbing, like with every other command, but double square brackets are special bash syntax, no command.
- As well try to avoid regex expanding `=~` where possible. `=~ ^-` equals `== -*`, but the letter is faster and less prone for accidents.
- On the right side of double square brackets, globs are however expended (but spaces taken literally), so there quotation makes sense to prevent globbing. Same as with command arguments, single quotes are to be preserved, if no `$` shall be expanded.

But all pretty minor of course.

One question: I saw that `2>>/dev/null` the second time now, while I use `2>/dev/null`. Do you know any difference this has with that particular meta file or is it just what you're used to?

ollliegits replied:
Thanks @MichaIng, for your review. I should be able to update the PR by tomorrow.

MichaIng replied:
Great, many thanks for that. Let me know if I can help.

MichaIng commented on node-install.sh:
```diff
@@ -1,6 +1,59 @@
+        elif command -v 'jq' &> /dev/null; then
+            curl -l "https://unofficial-builds.nodejs.org/download/release/index.json" 2>>/dev/null | jq -r '.[].version' | sort -Vr
+        else
+            echo "Required command line tools tail + cut, OR jq not found."
```
Let's stay consistent with error messages (same below lines 34 and 52):
```
echo 'ERROR: Required command line tools tail + cut, OR jq not found.' >&2
```

ollliegits replied:
Fixed.

MichaIng commented on node-install.sh:
```diff
@@ -16,9 +69,17 @@ if [[ $EUID != 0 ]]
-URL='https://nodejs.org/dist/'
-if [[ $ARCH == 'aarch64' ]]
+
+if [[ ! -z "$unofficial_build_version" ]]; then
```
`[[ -n $unofficial_build_version ]]` or simply `[[ $unofficial_build_version ]]`. No need for doubled negation, and the quotation here as no effect, same like with left sided strings 🙂.

ollliegits replied:
Fixed.

MichaIng commented:
The PR comes at a perfect time. We currently have an RPi Zero (ARMv6) user that has issues with Node-RED. And the reason seems to be Node v11, latest official ARMv6 build, but unofficials have v15.14, which should fix the issue 👍.

Commit:
fix: MichaIngs PR feedback applied.

Commit:
chore: Codestyle in curl URL.

ollliegits committed:
Merge pull request #3 from ollliegits/master

finalize armv6 unofficials

ollliegits commented:
Thank you very much for the great feedback and hints! Been fun.

MichaIng committed:
Minor alignments

+ Align and print error messages to STDOUT
+ Print script info and authors on every execution, not just as part of help screen.
+ Compact help screen to align with common Linux utilities
+ Error out when invalid input argument is given
+ Print curl errors when checking for unofficial versions and assure that HTTP error responses are not parsed but printed compact (response code) to STDERR.

MichaIng committed:
Add more info about unofficial builds

MichaIng commented:
I added another polish commit, I hope you don't mind. Preparing to add further parameter options soon. Running some tests now.

Commit:
feat: Support for retrieval of latest version of unofficial builds added.

MichaIng commented on node-install.sh:
```diff
@@ -1,6 +1,71 @@
+        shift; unofficial_build_version=$1
+        if [[ -z $unofficial_build_version ]]; then
+            unofficial_build_version=$(curl -sSf $UNOFFICIALS_URI/index.tab | grep $ARCH | cut -f1 | tail -n +2 | sort -V | tail -n 1)
+            echo "Found latest version $unofficial_build_version for architecture $ARCH."
```
As long as the official list is not included when using the `-u` flag, shall we add another `[[ -z $unofficial_build_version ]] && echo "ERROR: No version found in unofficial builds for architecture $ARCH" && exit 1`? E.g. when using the flag with a regular amd64 or armhf system. And of course theoretically there might be completely unsupported architectures, or the arch estimation fails.

MichaIng commented on node-install.sh:
```diff
@@ -1,6 +1,71 @@
+    -u | --unofficial )
+        shift; unofficial_build_version=$1
+        if [[ -z $unofficial_build_version ]]; then
+            unofficial_build_version=$(curl -sSf $UNOFFICIALS_URI/index.tab | grep $ARCH | cut -f1 | tail -n +2 | sort -V | tail -n 1)
```
Ah, as we sort out lines without matching architecture first (which includes the column names), `tail -n +2` removes that latest version.

MichaIng commented:
Works fine, but probably we can enhance the output a little:
```
Universal Node.js Linux Installer by github.com/taaem, updated by github.com/MichaIng
Found latest version v15.14.0 for architecture armv6l.
root permissions verified
Searching version v15.14.0 for armv6l ...
Found latest stable version for armv6l: node-v15.14.0-linux-armv6l.tar.gz
Downloading https://unofficial-builds.nodejs.org/download/release/v15.14.0/node-v15.14.0-linux-armv6l.tar.gz ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29.1M  100 29.1M    0     0  35179      0  0:14:28  0:14:28 --:--:-- 41962
Finished downloading!
Installing /tmp/node.tar.gz ...
removed '/tmp/node.tar.gz'
Finished installing!
```
- Shall we do the root permissions check before checking for unofficial versions? Either we can check for the latest unofficial version outside of the input arguments loop, or we simply do it before parsing input arguments at all? On the other hand, help screen and listing versions does not require root. ... okay let's keep it in mind but leave it as it is for now. When reworking official version detection+selection, this can be done.
- Let's move the latest stable version prints [line 118 - 124](https://github.com/MichaIng/nodejs-linux-installer/pull/2/files#diff-84d07f6d1c216dd9dbceedccbb6cbd1de08e2caa432cc550928e0251d2cd1a57R118-R124) into the official versions detection block, as it looks strange that first a latest version was found, then a latest stable version is found again.
- Also `Searching version v15.14.0 for armv6l ...` while before `Found latest version v15.14.0 for architecture armv6l.` looks strange. Let's change it to make clear that it's not about searching a version but checking if a download is available.

I'll do a commit to address this.

MichaIng committed:
Simplify getting unofficial versions + alignments

Since the index.tab is sorted already, we can simply pick the first match. Merge multiple tools into a single awk call.

Additionally order and outputs have been adjusted to be more reasonable, so that e.g. no version is searched before root permissions have been verified, and no "latest version found" is printed when a specific version was given. Also the doubled download check (HEAD request) for unofficial version has been removed, the actual download will fail instead, printing a meaningful curl error messages, in case.

Lastly the if-then-else blocks have been further aligned to match the most common code style.

MichaIng committed:
Syntax

MichaIng commented:
Okay, LGTM now:
```
# ./node-install.sh -u v12.22.1
Universal Node.js Linux Installer by github.com/taaem, updated by github.com/MichaIng
root permissions verified
The following unofficial build version was requested: v12.22.1
Downloading https://unofficial-builds.nodejs.org/download/release/v12.22.1/node-v12.22.1-linux-armv6l.tar.gz ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 20.3M  100 20.3M    0     0  1879k      0  0:00:11  0:00:11 --:--:-- 4152k
Finished downloading!
Installing /tmp/node.tar.gz ...
removed '/tmp/node.tar.gz'
Finished installing!

# ./node-install.sh -u
Universal Node.js Linux Installer by github.com/taaem, updated by github.com/MichaIng
root permissions verified
Searching latest unofficial build version for armv6l ...
Found latest unofficial build version for armv6l: v16.1.0
Downloading https://unofficial-builds.nodejs.org/download/release/v16.1.0/node-v16.1.0-linux-armv6l.tar.gz ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29.3M  100 29.3M    0     0  5355k      0  0:00:05  0:00:05 --:--:-- 6296k
Finished downloading!
Installing /tmp/node.tar.gz ...
removed '/tmp/node.tar.gz'
Finished installing!

# sed -i '/^ARCH=/c\ARCH=x86_64' ./node-install.sh

# ./node-install.sh -u
Universal Node.js Linux Installer by github.com/taaem, updated by github.com/MichaIng
root permissions verified
Searching latest unofficial build version for x86_64...
ERROR: Failed to find any unofficial build version for x86_64

# ./node-install.sh
Universal Node.js Linux Installer by github.com/taaem, updated by github.com/MichaIng
root permissions verified
Searching latest stable version for x86_64 ...
Found latest stable version for x86_64: node-v16.1.0-linux-x64.tar.gz
Downloading https://nodejs.org/dist/latest/node-v16.1.0-linux-x64.tar.gz ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 31.7M  100 31.7M    0     0  6450k      0  0:00:05  0:00:05 --:--:-- 6931k
Finished downloading!
Installing /tmp/node.tar.gz ...
removed '/tmp/node.tar.gz'
Finished installing!
```

MichaIng committed:
Minor

Not the version is unofficial, but the builds. Also add an additional echo when listing versions, as otherwise it looks like a bug when no version was found for the architecture.

ollliegits reviewed (commented):
Very nice, the code has gotten so much cleaner.

MichaIng commented:
Okay nice, let's get this merged.
